packages/fast-usdc/src/exos/settler.js

```diff
@@ -1,17 +1,95 @@
+import { assertAllDefined } from '@agoric/internal';
+import { atob } from '@endo/base64';
+import { makeError, q } from '@endo/errors';
+import { M } from '@endo/patterns';
+
+import { addressTools } from '../utils/address.js';
+
 /**
+ * @import { FungibleTokenPacketData } from '@agoric/cosmic-proto/ibc/applications/transfer/v2/packet.js';
+ * @import {Denom} from '@agoric/orchestration';
+ * @import {IBCChannelID, VTransferIBCEvent} from '@agoric/vats';
  * @import {Zone} from '@agoric/zone';
+ * @import { NobleAddress } from '../types.js';
  * @import {StatusManager} from './status-manager.js';
  */
 
-import { assertAllDefined } from '@agoric/internal';
-
 /**
  * @param {Zone} zone
  * @param {object} caps
  * @param {StatusManager} caps.statusManager
  */
 export const prepareSettler = (zone, { statusManager }) => {
   assertAllDefined({ statusManager });
-  return zone.exo('Fast USDC Settler', undefined, {});
+  return zone.exoClass(
+    'Fast USDC Settler',
+    M.interface('SettlerI', {
+      receiveUpcall: M.call(M.record()).returns(M.promise()),
+    }),
+    /**
+     *
+     * @param {{
+     *   sourceChannel: IBCChannelID;
+     *   remoteDenom: Denom
+     * }} config
+     */
+    config => harden(config),
+    {
+      /** @param {VTransferIBCEvent} event */
+      async receiveUpcall(event) {
+        if (event.packet.source_channel !== this.state.sourceChannel) {
+          // only interested in packets from the issuing chain
+          return;
+        }
+        const tx = /** @type {FungibleTokenPacketData} */ (
+          JSON.parse(atob(event.packet.data))
+        );
+        if (tx.denom !== this.state.remoteDenom) {
+          // only interested in uusdc
+          return;
+        }
+
+        if (!addressTools.hasQueryParams(tx.receiver)) {
+          // only interested in receivers with query params
+          return;
+        }
+
+        const { params } = addressTools.getQueryParams(tx.receiver);
+        // TODO - what's the schema address parameter schema for FUSDC?
+        if (!params?.EUD) {
+          // only interested in receivers with EUD parameter
+          return;
+        }
+
+        const hasPendingSettlement = statusManager.hasPendingSettlement(
+          tx.sender,
+          BigInt(tx.amount),
+        );
+        if (!hasPendingSettlement) {
+          // TODO FAILURE PATH -> put money in recovery account or .transfer to receiver
+          // TODO should we have a TxStatus for this?
+          throw makeError(
+            `🚨 No pending settlement found for ${q(tx.sender)} ${q(tx.amount)}`,
```

Comment:
Is this recoverable? It looks like it.

```suggestion
            `⚠️ No pending settlement found for ${q(tx.sender)} ${q(tx.amount)}`,
```

per [Logging docs](https://github.com/Agoric/agoric-sdk/wiki/Logging):
> 🚨 indicates something that should never happen.
>
> ⚠️ indicates something that is not expected but was recovered from.

Reply:
🚨 was intentional - I think we should leave it until we recover from here

Reply:
We might want to capture this in the StatusManager with an  `ORPHANED` state, and add some logic in the `observe()` method to initiate a `SETTLE` / settlement

Reply:
Let's come back to this when we're going all the error cases